configfile: "configs/config.yaml"


rule all:
    input:
        expand(config['short_rnaseq']['polyA_trimmed'],
               encode_id=config['short_rnaseq']['data']['gm12878'],
               sample='gm12878')


rule download_short_reads:
    output:
        config['short_rnaseq']['fastq']
    shell:
        "wget https://www.encodeproject.org/files/{wildcards.encode_id}/@@download/{wildcards.encode_id}.fastq.gz -O {output}"


rule filter_reads_with_polyA_tail:
    input:
        fastq = config['short_rnaseq']['fastq']
    params:
        pattern = "A" * 10,
        mismatch = 0
    output:
        fastq = config['short_rnaseq']['polyA_trimmed']
    shell:
        '''
        seqkit grep --by-seq --max-mismatch {params.mismatch} \
        --pattern {params.pattern} {input.fastq} | gzip > {output.fastq}
        '''


rule chrom_sizes:
    input:
        fasta = config['fasta']
    output:
        chrom_sizes = config['chrom_sizes']
    shell:
        "faidx {input.fasta} -i chromsizes > {output.chrom_sizes}"

rule postprocess_tes:
    input:
        read_annot = config['pilot']['read_annot'],
        fasta = config['fasta'],
        gtf = config['gtf'],
        chrom_sizes = config['chrom_sizes']
    output:
        directory(config['pilot_apa']['dir'])
    shell:
        "lapa \
        --alignment {input.read_annot} \
        --fasta {input.fasta} \
        --annotation {input.gtf} \
        --chrom_sizes {input.chrom_sizes} \
        --output_dir {output}"


rule pilot_plot:
    input:
        read_annot = config['pilot']['read_annot'],
        fasta = config['fasta'],
        gtf = config['gtf'],
        chrom_sizes = config['chrom_sizes']
    output:
        directory(config['pilot_apa']['dir'])
    shell:
        "lapa \
        --alignment {input.read_annot} \
        --fasta {input.fasta} \
        --annotation {input.gtf} \
        --chrom_sizes {input.chrom_sizes} \
        --output_dir {output}"
