import pandas as pd

configfile: "configs/config.yaml"


df_lrgasp = pd.read_csv(config['lrgasp']['config'], sep='\t')


def genome(wildcards):
    if 'genome' in wildcards.keys():
        return wildcards.genome

    if wildcards.sample == 'ES':
        return 'mm39'
    elif wildcards.sample == 'WTC11':
        return 'hg38'
    # if wildcards.sample == 'gm12878':
    #     return 'human'
    # elif wildcards.sample == 'c2c12':
    #     return 'mouse'
    else:
        raise ValueError(f'{wildcards.sample} is not known sample')


def fasta(wildcards):
    return config['genome']['fasta'][genome(wildcards)]


def gtf(wildcards):
    return config['genome']['gtf'][genome(wildcards)]


def chrom_sizes(wildcards):
    return config['common']['chrom_sizes'].format(genome=genome(wildcards))


def star_index(wildcards):
    return config['genome']['star_index'][genome(wildcards)]


# df_lrgasp = df_lrgasp[
#     (df_lrgasp['file_contents'] == 'reads') &
#     (df_lrgasp['library_prep'] == 'dRNA') &
#     (df_lrgasp['sample'] == 'WTC11')
# ]


# def read_annot(wildcards):
#     if wildcards.sample == 'gm12878':
#         return config['pilot']['read_annot']
#     elif wildcards.sample == 'c2c12':
#         return config['c2c12']['read_annot_bulk']
#     else:
#         raise('readannot is not defined for `%s`' % wildcards.sample)

include: "common/Snakefile"
include: "long_reads/Snakefile"
include: "align_short_rnaseq/Snakefile"
include: "benchmark/Snakefile"
include: "correction/Snakefile"
include: "wg_encode/Snakefile"

# include: "quantseq3/Snakefile"
# include: "wtc11/Snakefile"
# include: "read3/Snakefile"
# include: "c2c12/Snakefile"
# include: "pilot/Snakefile"

pb_library_prep = ['cDNA', 'CapTrap']
ont_library_prep = ['cDNA', 'dRNA', 'CapTrap', 'R2C2']


rule all:
    input:
        config['wg_encode']['human']['read_annot'],
        config['wg_encode']['human']['gtf'],
        config['wg_encode']['human']['abundance'],
        config['wg_encode']['mouse']['read_annot'],
        config['wg_encode']['mouse']['gtf'],
        config['wg_encode']['mouse']['abundance']


        # expand(rules.plot_overlap_annotation.output,
        #        sample='WTC11', library_prep='cDNA',
        #        platform='PacBio', counting='end')

        # expand(config['talon']['gtf_corrected'],
        #        sample='WTC11', library_prep='cDNA',
        #        platform='PacBio', counting='end')

#        expand(rules.plot_overlap_annotation.output,
#                sample='WTC11', library_prep='dRNA', platform='ONT'),
#         expand(rules.plot_overlap_corrected_annotation.output,
#                sample='WTC11', library_prep='dRNA',
#                platform='ONT', counting='end')
# -

    # expand(config['short_rnaseq']['bam'],
    #        encode_id=config['short_rnaseq']['wtc11']['se'], sample='WTC11')
    # expand(config['lapa']['lapa_dir'], sample='WTC11', platform=['PacBio'],
    #        counting=['end', 'tail'], library_prep=pb_library_prep),
    # expand(config['lapa']['lapa_dir'], sample='WTC11', platform=['ONT'],
    #        counting=['end', 'tail'], library_prep=ont_library_prep)
