import pdb
configfile: "configs/config.yaml"


def get_specie(wildcards):
    if 'specie' in wildcards.keys():
        return wildcards.specie

    if wildcards.sample == 'gm12878':
        return 'human'
    elif wildcards.sample == 'c2c12':
        return 'mouse'
    else:
        raise ValueError(f'{sample} is not known sample')


def fasta(wildcards):
    return config[get_specie(wildcards)]['fasta']


def gtf(wildcards):
    return config[get_specie(wildcards)]['gtf']


def chrom_sizes(wildcards):
    return config['common']['chrom_sizes'].format(specie=get_specie(wildcards))


def read_annot(wildcards):
    if wildcards.sample == 'gm12878':
        return config['pilot']['read_annot']
    elif wildcards.sample == 'c2c12':
        return config['c2c12']['read_annot_bulk']
    else:
        raise('readannot is not defined for `%s`' % wildcards.sample)


include: "common/Snakefile"
include: "read3/Snakefile"
include: "quantseq3/Snakefile"
include: "align_short_rnaseq/Snakefile"
include: "gm12878/Snakefile"
include: "c2c12/Snakefile"
include: "wtc11/Snakefile"
include: "benchmark/Snakefile"
include: "plot/Snakefile"


rule all:
    input:
        # figure-1
        rules.read_nums_datasets.output,
        rules.plot_hist_read_count.output,
        rules.site_motif.output,

        expand('data/results/gm12878/overlap_cage/{sample}_overlap_{platform}_annotation.csv',
               sample='gm12878', platform=['pb']),

        expand('reports/figures/{sample}_overlap_{platform}_annotation.png',
               sample='gm12878', platform=['pb', 'ont', 'pb_cor', 'ont_cor']),

        expand('reports/figures/{sample}_overlap_cage_{platform}_annotation.png',
               sample='gm12878', platform=['pb', 'ont', 'pb_cor', 'ont_cor']),
        [
            expand(
                'reports/figures/overlap/{tool}_{library_prep}_{platform}_{method}_annotation_cage_wtc11.png',
                tool=['talon', 'talon_lapa'],
                platform=platform,
                library_prep=config['long_read']['wtc11']['platform'][platform],
                method=['long'],  # , 'short'
                sample='wtc11'
            )
            for platform in config['long_read']['wtc11']['platform']
        ],
        [
            expand(
                'reports/figures/overlap/{tool}_{library_prep}_{platform}_{method}_annotation_wtc11.png',
                tool=['talon', 'talon_lapa'],
                platform=platform,
                library_prep=config['long_read']['wtc11']['platform'][platform],
                method=['long'],  # , 'short'
                sample='wtc11'
            )
            for platform in config['long_read']['wtc11']['platform']
        ]

        # config['long_read']['gm12878']['gtf_pb_cor']

        # expand(rules.count_gm12_pacbio.output,
        #        encode_id=config['long_read']['gm12878']['encode_id'])

        # expand(rules.merge_short_reads.output, sample='gm12878')

        # expand(config['lapa']['quantseq_dir'], sample=['gm12878']),
        # expand(config['lapa']['long_rna_dir'], sample=['gm12878']),
        # expand(config['lapa']['short_rna_dir'], sample=['gm12878']),
        # expand(config['lapa']['read3_dir'], sample=['c2c12']),
        # expand(config['lapa']['long_rna_dir'], sample=['c2c12']),

        # config['c2c12']['read_annot_bulk']
