import pytest


gtf_brca1 = 'tests/data/brca1.gtf'

nbr1_hepg2_read_annot = 'tests/data/nbr1_hepg2_read_annot.tsv'
read_annot = 'tests/data/chr17_read_annot.tsv'
fasta = 'tests/data/chr17.fa'
chr17_chrom_sizes = 'tests/data/chr17.chrom_sizes'
gtf = 'tests/data/chr17.gtf'
chrom_sizes = 'tests/data/chrom_sizes'

quantseq_gm12_bam = 'tests/data/quantseq3_gm12878_chr17_rep1.bam'
quantseq_both_gm12_bam = 'tests/data/quantseq3_gm12878_chr17_rep1.bam,tests/data/quantseq3_gm12878_chr17_rep2.bam'

sample_csv = 'tests/data/alignment_sample.csv'

agg_annotation_gene = [
    'tests/data/agg_gene_0.csv',
    'tests/data/agg_gene_1.csv',
    'tests/data/agg_gene_brca1.csv'
]

short_bam = 'tests/data/short_chr17.bam'
gtf_gm12_pb = 'tests/data/chr17_gm12_pb_talon.gtf'
read_annot_gm12_pb = 'tests/data/chr17_gm12_read_annot.tsv'
