
rule lapa_longreads:
    input:
        read_annot = read_annot,
        fasta = fasta,
        gtf = gtf,
        chrom_sizes = chrom_sizes
    threads: 1
    resources:
        mem_gb = 16
    output:
        config['lapa']['long_rna_dir']
    shell:
        "lapa \
        --alignment {input.read_annot} \
        --fasta {input.fasta} \
        --annotation {input.gtf} \
        --chrom_sizes {input.chrom_sizes} \
        --output_dir {output}"


rule lapa_gm12878_short_reads:
    input:
        bam = config['short_rnaseq']['merged_bam'],
        fasta = fasta,
        gtf = gtf,
        chrom_sizes = chrom_sizes
    threads: 1
    resources:
        mem_gb = 16
    output:
        directory(config['lapa']['short_rna_dir'])
    shell:
        "lapa \
        --alignment {input.bam} \
        --fasta {input.fasta} \
        --annotation {input.gtf} \
        --chrom_sizes {input.chrom_sizes} \
        --output_dir {output}"


rule sample_gm12878_quantseq:
    output:
        csv = config['lapa']['samples'].format(
            sample='gm12878', method='quantseq')
    run:
        bam = config['quantseq']['bam']
        with open(output['csv'], 'w') as f:
            f.write('sample,path\n')
            for i in ['rep1', 'rep2']:
                f.write(f'gm12878_{i},{bam.format(rep=i)}\n')


rule lapa_gm12878_quantseq:
    input:
        bam = expand(config['quantseq']['bam'], rep=['rep1', 'rep2']),
        sample_csv = config['lapa']['samples'].format(
            sample='gm12878', method='quantseq'),
        fasta = fasta,
        gtf = gtf,
        chrom_sizes = chrom_sizes
    threads: 1
    resources:
        mem_gb = 16
    output:
        directory(config['lapa']['quantseq_dir'])
    shell:
        "lapa \
        --alignment {input.sample_csv} \
        --fasta {input.fasta} \
        --annotation {input.gtf} \
        --chrom_sizes {input.chrom_sizes} \
        --output_dir {output}"
