
# rule sample_c2c12_read3:
#     output:
#         csv = config['lapa']['samples'].format(
#             sample='c2c12', method='read3')
#     run:
#         bam = config['c2c12']['read3']['bam']

#         with open(output['csv'], 'w') as f:
#             f.write('sample,path\n')
#             for i in config['c2c12']['read3']['rep']:
#                 f.write(f'c2c12_{i},{bam.format(rep=i)}\n')


# rule lapa_c2c12_read3:
#     input:
#         bam = expand(config['c2c12']['read3']['bam'],
#                      rep=config['c2c12']['read3']['rep']),
#         sample_csv = config['lapa']['samples'].format(
#             method='read3', sample='{sample}'),
#         fasta = fasta,
#         gtf = gtf,
#         chrom_sizes = chrom_sizes
#     threads: 1
#     resources:
#         mem_gb = 16
#     output:
#         directory(config['lapa']['read3_dir'])
#     shell:
#         "lapa \
#         --alignment {input.sample_csv} \
#         --fasta {input.fasta} \
#         --annotation {input.gtf} \
#         --chrom_sizes {input.chrom_sizes} \
#         --output_dir {output}"


rule split_c2c12_read_annot:
    input:
        read_annot = config['c2c12']['read_annot_bulk_sc']
    output:
        bulk_annot = config['c2c12']['read_annot_bulk'],
        sc_annot = config['c2c12']['read_annot_sc']
    script:
        "split_annot.py"


rule lapa_longreads_c2c12:
    input:
        aligment = config['c2c12']['read_annot_bulk'],
        fasta = config['mouse']['fasta'],
        gtf = config['mouse']['gtf'],
        chrom_sizes = config['common']['chrom_sizes'].format(specie='mouse')
    threads: 1
    resources:
        mem_gb = 16
    output:
        directory(config['lapa']['long_rna_dir'].format(sample='c2c12'))
    shell:
        "lapa \
        --alignment {input.aligment} \
        --fasta {input.fasta} \
        --annotation {input.gtf} \
        --chrom_sizes {input.chrom_sizes} \
        --method end \
        --output_dir {output}"


rule differential_apa_c2c12:
    input:
        long_read = config['lapa']['long_rna_dir'].format(sample='c2c12'),
        # quantseq = config['lapa']['quantseq_dir'].format(sample='c2c12'),
        mapping = config['gene_id_name_mapping'].format(specie='mouse')
    threads: 1
    resources:
        mem_gb = 16
    output:
        volcona_plot = '../../reports/figures/c2c12_volcona.png',
        heatmap_plot = '../../reports/figures/c2c12_heatmap.png',
        sig_csv = config['c2c12']['sig']
    notebook:
        "./differential_apa.ipynb"


rule differential_apa_abundance_c2c12:
    input:
        sig_csv = config['c2c12']['sig'],
        abundance = config['c2c12']['abundance']
    threads: 1
    resources:
        mem_gb = 16
    output:
        boxplot = '../../reports/figures/c2c12_boxplot_gene.png',
    notebook:
        "./apa_differetial_gene_exp.ipynb"

# rule lapa_c2c12_longreads:
#     input:
#         read_annot = config['c2c12']['read_annot_bulk'],
#         fasta = fasta,
#         gtf = gtf,
#         chrom_sizes = chrom_sizes
#     threads: 1
#     resources:
#         mem_gb = 16
#     output:
#         directory(config['lapa']['long_rna_dir'].format(sample='c2c12'))
#     shell:
#         "lapa \
#         --alignment {input.read_annot} \
#         --fasta {input.fasta} \
#         --annotation {input.gtf} \
#         --chrom_sizes {input.chrom_sizes} \
#         --output_dir {output}"
